internal: consistently use the controller-runtime informer cache

jpeach opened the pull request:
Prior to this change, the Kubernetes client API exposed methods to
create a informer factories as well as namespaced informer factories.
This encouraged other systems to use multiple factories and multiple
informers for each factory. The result of this is that objects could be
cached multiple times in the different informer factory caches, and that
the informer states could be subtly out of sync.

This change removes the internal APIs that expose informer factories.
There is now just one API to create an informer, and it is backed by the
controller-runtime cache API. This ensures that all the informers used
across Contour share a common cache and therefore a consistent view of
Kubernetes state.

This updates #2683.
This fixes #2904.

Signed-off-by: James Peach <jpeach@vmware.com>

jpeach commented:
Need to test with knative integration tests tomorrow.

skriss commented on cmd/contour/serve.go:
```diff
@@ -374,50 +368,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-		} else {
-			informerSyncList.InformOnResources(clusterInformerFactory, dynamicHandler, k8s.ServiceAPIResources()...)
+		for _, r := range k8s.ServiceAPIResources() {
+			if !clients.ResourcesExist(k8s.ServiceAPIResources()...) {
```
```suggestion
			if !clients.ResourcesExist(r) {
```

skriss commented on cmd/contour/serve.go:
```diff
@@ -374,50 +368,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+	g.Add(func(stop <-chan struct{}) error {
+		log := log.WithField("context", "informers")
+
+		log.Info("started informers")
```
should either be "starting", or move to after they've actually been started

skriss commented on cmd/contour/ingressstatus.go:
```diff
@@ -14,32 +14,35 @@
 // The theory of operation of the loadBalancerStatusWriter is as follows:
+//
 // 1. On startup the loadBalancerStatusWriter waits to be elected leader.
 // 2. Once elected leader, the loadBalancerStatusWriter waits to receive a
```
it'd be helpful to also document that once elected leader, it creates a `StatusAddressUpdater` which will handle writing LBStatus for future Ingress/HTTPProxy objects.

jpeach replied:
I added this info, though in a slightly different place.

skriss commented on cmd/contour/serve.go:
```diff
@@ -215,22 +213,18 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	if rootNamespaces := ctx.proxyRootNamespaces(); len(rootNamespaces) > 0 {
+		informerNamespaces = append(informerNamespaces, rootNamespaces...)
+
 		// Add the FallbackCertificateNamespace to the root-namespaces if not already
```
```suggestion
		// Add the FallbackCertificateNamespace to informerNamespaces if not already
```

skriss reviewed (changes requested):
Handful of comments but this looks pretty good and sets us up for using the controller-runtime cache in place of our hand-rolled one. I do wonder if there are additional places where we need to wait for the cache to sync during startup (AFAIK right now we only do it before starting the xds server), but i'll need to look some more to see where that might be.

jpeach commented:
Seeing this error a couple of times when I run the Knative integration tests:
```
time="2020-10-07T08:39:31Z" level=error msg="unable to update status" context=StatusUpdateHandler error="Operation cannot be fulfilled on httpproxies.projectcontour.io \"ingress-conformance-0-dispatch-ede9da76e278bca44bfaa9dee298cf80\": StorageError: invalid object, Code: 4, Key: /registry/projectcontour.io/httpproxies/serving-tests/ingress-conformance-0-dispatch-ede9da76e278bca44bfaa9dee298cf80, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: c53843a2-7db7-4343-97fa-c1dd1dd47380, UID in object meta: " name=ingress-conformance-0-dispatch-ede9da76e278bca44bfaa9dee298cf80 namespace=serving-tests
```

jpeach commented:
@skriss Could you please take another look. I ran into Go module hell (which is why I had to update the indirect `go-logr/logr` dependency) with knative, so going to have to table that stuff for a while. I think I've addressed all your feedback, and I added conflict retries to take care of the conflict error I saw logged.

skriss commented on cmd/contour/ingressstatus.go:
```diff
@@ -62,62 +65,76 @@ func (isw *loadBalancerStatusWriter) Start(stop <-chan struct{}) error {
+		Converter:     isw.Converter,
+	}
+
+	// Create new informer for the new LoadBalancerStatus. The
```
nit: this comment is slightly stale, could be clarified since we're no longer creating a new informer every time the LB address changes.

jpeach replied:
The comment doesn't say that it is creating a new informer every time?

skriss replied:
it says "for the new LoadBalancerStatus", which came from the old code where when a new LB address was received, a new informer would be created. Can we just say "Create informers" or something similar that doesn't reference the old code structure?

jpeach replied:
Ah I see ... fixed.

skriss commented on internal/k8s/status.go:
```diff
@@ -60,19 +62,71 @@ func (m StatusMutatorFunc) Mutate(old interface{}) interface{} {
+		return
+	}
+
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
this looks reasonable to me. Were the conflicts due to Contour writing both conditions updates and LB address updates ~concurrently?

jpeach replied:
I wasn't able to debug it. Rebasing onto the Kubernetes 1.19 Go client broke knative with module dependencies :(

skriss commented on cmd/contour/serve.go:
```diff
@@ -374,50 +369,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+	})
 
 	// Register our event handler with the workgroup.
 	g.Add(eventHandler.Start())
```
I *think* that we should probably be waiting for the informer caches to sync in here before starting to process events to avoid any race conditions with the caches populating, but I'd have to experiment a bit to confirm if needed/not problematic.

jpeach replied:
Can you elaborate why you think this is needed? This change keeps the current semantics of not starting the xDS server until the informer cache is present. That's not awesome, but it's not different than what we had.

jpeach replied:
https://github.com/projectcontour/contour/issues/1280

skriss replied:
Typically in Kubernetes controllers, you would wait for the informer caches to sync (via WaitForCacheSync) before starting to process events from those informers. I think the reason for this is that if you don't, then you may get not-found errors if you try to look up other items from the informer cache as part of your event handler logic.

One scenario I'm thinking of is in the load balancer address update code, which handles an event for the load balancer service, then tries to look up all ingresses/proxies from the cache so it can write addresses to them. if the ingress/informer caches haven't fully synced yet, then we could get an incomplete list back and not update all of them as expected. 

Practically it may or may not be an issue for us, I'm not sure. If we're not seeing issues, then we can defer/ignore.

jpeach replied:
> One scenario I'm thinking of is in the load balancer address update code, which handles an event for the load balancer service, then tries to look up all ingresses/proxies from the cache so it can write addresses to them. if the ingress/informer caches haven't fully synced yet, then we could get an incomplete list back and not update all of them as expected.

I don't think this is a problem. The List is for updating objects that won't get passed to `OnAdd`. If the cache isn't synced when the List happens, then the `OnAdd` would work.

stevesloka replied:
The current code does wait for the caches to be synced, but doesn't wait for the first dag rebuild to happen before starting the xds server. Need to make sure that behavior is still kept. Contour needs to wait for the caches to be synced before starting things up. The 2nd bit is still an open issue that needs solved.

jpeach replied:
IIUC, we all agree that this PR doesn't change the current Contour semantics (even though those semantics leave something to be desired)? If so, then we can resolve this thread.

skriss replied:
filed #3020

skriss commented on cmd/contour/serve.go:
```diff
@@ -503,7 +527,10 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	g.Add(lbsw.Start)
```
similar to the main contour event handler, I *think* that we should probably be waiting for the informer caches to sync in here before starting to process events to avoid any race conditions with the caches populating, but I'd have to experiment a bit to confirm if needed/not problematic.

jpeach replied:
???

jpeach replied:
Yeh, that's a different issue though.

skriss replied:
filed #3020

skriss reviewed (changes requested):
changes look pretty good to me. I added thoughts on where we might want to wait for caches to sync before processing events, but as I noted, I'd have to experiment to confirm if that's actually needed, and doesn't break anything.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -511,24 +538,31 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-		informerSyncList.InformOnResources(factory, dynamicServiceHandler, k8s.ServicesResources()...)
 
-		g.Add(startInformer(factory, log.WithField("context", "serviceStatusLoadBalancerWatcher")))
+		for _, r := range k8s.ServicesResources() {
```
Might help for readability to move this into a method to clean up the serve method.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -372,50 +367,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-	informerSyncList.InformOnResources(clusterInformerFactory,
-		&k8s.DynamicClientHandler{
+	// Inform on endpoints.
+	for _, r := range k8s.EndpointsResources() {
```
Might help for readability to move this into a method to clean up the serve method.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -372,50 +367,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-		informerSyncList.InformOnResources(factory, dynamicHandler, k8s.SecretsResources()...)
-	}
+	// Inform on secrets, filtering by root namespaces.
+	for _, r := range k8s.SecretsResources() {
```
Might help for readability to move this into a method to clean up the serve method.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -372,50 +367,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-			log.WithField("InformOnResources", "ExperimentalServiceAPITypes").Warnf("resources %v not found in api server", k8s.ServiceAPIResources())
-		} else {
-			informerSyncList.InformOnResources(clusterInformerFactory, dynamicHandler, k8s.ServiceAPIResources()...)
+		for _, r := range k8s.ServiceAPIResources() {
```
Might help for readability to move this into a method to clean up the serve method.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -511,24 +538,31 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-		if err := informerSyncList.WaitForSync(stop); err != nil {
-			return err
+		if !clients.WaitForCacheSync(stop) {
+			return errors.New("informer cache failed to synx")
```
s/synx/sync

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -372,50 +367,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
-	// If root namespaces are not defined, then add the informer for all namespaces
-	if len(namespacedInformerFactories) == 0 {
-		informerSyncList.InformOnResources(clusterInformerFactory, dynamicHandler, k8s.SecretsResources()...)
+		if len(informerNamespaces) == 0 {
```
Add the comment back describing this if statement.

jpeach replied:
Fixed.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -501,7 +525,10 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	// Register an informer to watch envoy's service if we haven't been given static details.
-	if ctx.IngressStatusAddress == "" {
+	if ctx.IngressStatusAddress != "" {
+		log.WithField("loadbalancer-address", ctx.IngressStatusAddress).Info("Using supplied information for Ingress status")
```
s/loadbalancer-address/ingress-status-address to keep consistent with the config name.

jpeach replied:
This field was present in the original code (the log line must moved). We should keep this field name because it correlates with the log message in `serve.go`.

stevesloka replied:
I don't agree, Contour should be consistent with its logs and matching the same name against what it's logging is what we should strive to do.

jpeach replied:
This is consistent. You need to think about logs from the perspective of someone trying to triage a problem in production. That person doesn't know what the internal names are in the code, but can correlate context tags if we use them consistently.

When we use different tag names, we make life harder for operators. If we apply your suggestion, then there are different context tags for setting the address in config vs. applying the address to an object. That makes is unnecessarily hard for operators to understand what Contour is doing with the address.

At any rate, this code has not changed in this PR.

This PR:
```
cmd/contour/ingressstatus.go:                   isw.log.WithField("loadbalancer-address", lbAddress(lbs)).
cmd/contour/serve.go:           log.WithField("loadbalancer-address", ctx.IngressStatusAddress).Info("Using supplied information for Ingress status")
```

main:
```
cmd/contour/ingressstatus.go:                   isw.log.WithField("loadbalancer-address", lbAddress(lbs)).Info("received a new address for status.loadBalancer")
cmd/contour/serve.go:           log.WithField("loadbalancer-address", ctx.IngressStatusAddress).Info("Using supplied information for Ingress status")
```

stevesloka replied:
Right, to be consistent we should use the name that operators use to configure the address. When specifying an address they set `ingress-status-address`, not `loadbalancer-address` which is what is log is referencing.

stevesloka replied:
If you don't want to change the the log line @jpeach I'll do a follow up PR to make it consistent.

jpeach replied:
This patch doesn't change the existing code. I think that if we want to make that change, it would be best to have a separate issue.
